### PR TITLE
Added confirmation prompt for sky storage delete, and --yes flag to skip it

### PR DESCRIPTION
**Pull request by aseriesof-tubes**
This is for issue #2635 

Added a feature for `sky storage delete` where by default it shows a confirmation prompt before deleting, like in `sky down` or `sky stop`. Like those commands, users can add a `--yes` flag to skip the confirmation prompt. Also changed some tests in test_smoke for TestStorageWithCredentials for the tests to automatically use the --yes flag. 

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
Changed some `check_output` commands to include the `--yes` flag, so previously if it was `check_output(
            ['sky', 'storage', 'delete', storage_name])`, now it's `check_output(['sky', 'storage', 'delete', storage_name, '--yes'])`   
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [x] Relevant individual smoke tests: `pytest tests/test_smoke.py::TestStorageWithCredentials` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/cli.py
Comment (landscapepainter) [inline sky/cli.py:3478]:
```suggestion
        sky storage delete imagenet -y
```
We need to provide a storage name to delete. Using a same storage as the example given above.

File: sky/cli.py
Comment (landscapepainter) [inline sky/cli.py:3506]: There are two major components to be achieved in this snippet from `L3482`-`L3512`. 

1. obtain the name of the storages to be deleted
2. Prompt the user if `-y` is not passed.

And these can be achieved with simpler manner if we can reorder the codes and dynamically set the message of the prompt. We can simply add a prompt once at `L3513` after obtaining the names of the storages to be deleted either with `all` or not. Then, to prompt, use something like the following after renaming the variable `names` to `storage_list`:
```
    if not yes:
        storage_names = ', '.join(storage_list)
        storage_str = 'storages' if len(storage_list) > 1 else 'storage'
        click.confirm(
            f'Deleting {len(storage_list)} {storage_str}: '
            f'{storage_names}. Proceed?',
            default=True,
            abort=True,
            show_default=True)
```

**Review by landscapepainter — COMMENTED** [review]
@aseriesof-tubes Thanks for your contribution! Left some comments :)

File: sky/cli.py
Comment (aseriesof-tubes) [reply sky/cli.py]: This has been fixed!

File: sky/cli.py
Comment (romilbhardwaj) [inline sky/cli.py:3496]: When I run `sky storage delete -a`, a redundant `Deleting all storage objects.` is printed.
```
(base) ➜  sky-experiments git:(issue2708) ✗ sky storage delete -a             
Deleting all storage objects.
Deleting 1 storage: romil-test-bucket. Proceed? [Y/n]: ^CAborted!
```

Can we instead just show:
```
(base) ➜  sky-experiments git:(issue2708) ✗ sky storage delete -a             
Deleting <count> storages: romil-test-bucket, ... . Proceed? [Y/n]:
```

Comment (romilbhardwaj) [inline sky/cli.py:3496]: We should exit if no storages are found, but a prompt is shown:
```
(base) ➜  ~ sky storage delete abc
Storage abc not found.
Deleting 0 storage: . Proceed? [Y/n]: 
```

Comment (romilbhardwaj) [reply sky/cli.py:3496]: Same when a storage name is specified:
```
(base) ➜  ~ sky storage delete romil-test-bucket
Deleting 1 storage object.
Deleting 1 storage: romil-test-bucket. Proceed? [Y/n]:
```

**Review by romilbhardwaj — COMMENTED** [review]
Welcome to SkyPilot @aseriesof-tubes! I noticed some redundant messages and a bug when non-existent storage is specified, noted in comments below.

File: sky/cli.py
Comment (aseriesof-tubes) [reply sky/cli.py:3478]: Yes, now that I see it, the additional message isn't consistent with the other cli messages, will be fixed! Thank you for pointing that out.

File: sky/cli.py
Comment (aseriesof-tubes) [reply sky/cli.py:3496]: This has been fixed, and `_get_glob_storages` has been modified accordingly.

Comment (aseriesof-tubes) [reply sky/cli.py:3496]: I've changed the `if not yes` if statement to check for `names` if it has length zero. the subprocess util should take care of the rest if nothing is found.

**Comment by aseriesof-tubes** [comment]
Thank you for the warm welcome @romilbhardwaj and @landscapepainter, your reviews are appreciated, and changes have been made!

File: sky/cli.py
Comment (landscapepainter) [inline sky/cli.py:3481]: Some edge case to keep in mind while resolving the comment above is to make sure to handle the case when multiple storage names are given, but some of them do not exist. For example, when you have following storages, and when `sky storage delete doyoung-r2-test storage-not-exsit` is ran, we want to make sure to inform that `storage-not-exist` does not exist and prompt the user for deleting the existing ones, which in this case is `doyoung-r2-test`. This is currently handled well by `_get_glob_storages`, but still wanted to give some heads up.

```
$ sky storage ls
NAME                            UPDATED      STORE      COMMAND                       STATUS  
doyoung-r2-test                 12 mins ago  R2         sky launch yaml/r2_s3_sto...  READY   
doyoung-s3-test                 12 mins ago  S3         sky launch yaml/r2_s3_sto...  READY
$ sky storage delete doyoung-r2-test storage-not-exist
Storage storage-not-exist not found.
Deleting 1 storage: doyoung-r2-test. Proceed? [Y/n]:
```

File: tests/test_smoke.py
Comment (landscapepainter) [inline tests/test_smoke.py:3099]: nit: for conciseness and conveying the intention more clearly.
```suggestion
        delete_cmd = ['sky', 'storage', 'delete', '--yes']
        delete_cmd += storage_obj_name
```

**Review by landscapepainter — COMMENTED** [review]
Thanks for theq quick update @aseriesof-tubes ! Left some comments. This should be ready to merge after 👍

File: sky/cli.py
Comment (aseriesof-tubes) [reply sky/cli.py:3481]: Thanks for pointing that out, I didn't realize it before :)

**Comment by aseriesof-tubes** [comment]
Hi @landscapepainter @romilbhardwaj, the PR is ready for review again! Thank you for your patience :)

File: sky/cli.py
Comment (landscapepainter) [inline sky/cli.py:3491]:
```suggestion
    if names:
        if not yes:
            storage_names = ', '.join(names)
            storage_str = 'storages' if len(names) > 1 else 'storage'
            click.confirm(
                f'Deleting {len(names)} {storage_str}: '
                f'{storage_names}. Proceed?',
                default=True,
                abort=True,
                show_default=True)
    elif all:
        click.echo('No storage(s) to delete.')
```
nit: refactored the conditions for readability and efficiency removing redundant `len(names)` check

**Review by landscapepainter — APPROVED** [review]
LGTM! Thanks for the contribution @aseriesof-tubes

File: sky/cli.py
Comment (aseriesof-tubes) [reply sky/cli.py:3491]: Great, thank you so much for the review!

File: sky/cli.py
Comment (landscapepainter) [inline sky/cli.py:3492]: @aseriesof-tubes Sorry I missed to mention one last fix. Let's move this check under L3477. We can check if `storages` is empty or not and echo `'No storage(s) to delete.'` and then `return`. When storage names are specified for `sky storage delete`, it's echoed in `_get_glob_storages`, and we want to handle the echo under the same logical block of `if else`.

Comment (aseriesof-tubes) [reply sky/cli.py]: No problem! I agree that would be a cleaner approach. All changes have been made, and it is now ready for another review!

**Pull request merged**
